Wip master

ezmiller opened the pull request:
@AnnJohn and @thomastuts: This lesson on presentation components is ready for review.

ezmiller committed:
Update .learn file

ezmiller committed:
Add first draft of README

ezmiller committed:
Add minor edits

ezmiller committed:
Add more minor edits

ezmiller committed:
Add minor edit

ezmiller committed:
Add minor word change

ezmiller committed:
Add sentence about interactivity

ezmiller committed:
Update section title

ezmiller committed:
Add reference to lesson pure functions

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+"referentially transparent"  function.  Our UI has become just a bit more predictable.
+And, as web developers who've worked on the front-end, we know what a boon that
+is, don't we? (To review pure functions at greater length, see
+[this lesson](https://github.com/learn-co-curriculum/javascript-pure-functions)
```
We generally don't refer to past or present lessons, (as in 'in our last lesson, ...'), but I think explicitly linking to the lesson might work. @AnnJohn is this something that's allowed? If so, do we need to reference the learn.co link instead of the GitHub link?

AnnJohn replied:
@thomastuts @ezmiller Yes, avoiding phrases like "last lesson" and trying to keep the lessons as modular as possible is ideal. But linking to other lessons is fine via Github. It's a bit wonky with the Learn link

ezmiller replied:
@AnnJohn @thomastuts sounds like then the reference to the other lesson is a good idea here and doesn't break the modularity since it doesn't say "last lesson"? Or should I remove the reference? What is the "Learn Link"?

thomastuts replied:
The GitHub link will do just fine. The 'Learn link' is a URL that refers to the lesson on the learn.co platform rather than GitHub, since most lessons will be read this way by students (I _think_). But again, GitHub link is fine!

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
 
-## Resources
+The predictability comes from the fact -- and here we can see the influence on
+React of the principles of functional programming -- that this function will
```
'and here we can see the influence on React of the principles of functional programming' -> 'and here we can see the influence of functional programming on React'? Just a suggestion, feels like this improves the readability

ezmiller replied:
Indeed. Done.

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+	return <input
+					 className="field field-light"
+					 onChange={this.props.onChange || noop}
+					 limit={this.props.limit || defaultLimit} />
```
This indentation seems a little borked, the other code samples have 2-space indentation (which is great)

thomastuts commented on README.md:
````diff
@@ -1,55 +1,185 @@
+So what's this look like? Here's our `TextField` component rendered as a so-called
+`functional stateless` component (a feature available in React since v0.14):
+
+```
````
Missing a language identifier here (`js` or `javascript`)

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+&mdash; -- one that takes an input and returns a (portion of) the UI.
+
+So what's this look like? Here's our `TextField` component rendered as a so-called
+`functional stateless` component (a feature available in React since v0.14):
```
Not sure if backticks are appropriate here

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+(probably) does not have state? Well, if in fact we can create a component
+that has no state, that means that our component doesn't even really need to be
+a Javascript object of type `Component` at all. It can just be a simple function
+&mdash; -- one that takes an input and returns a (portion of) the UI.
```
This line has two em-dashes

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+our checklist for the presentational component pattern was that the component
+(probably) does not have state? Well, if in fact we can create a component
+that has no state, that means that our component doesn't even really need to be
+a Javascript object of type `Component` at all. It can just be a simple function
```
Javascript -> JavaScript :neckbeard:

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+stateless, all of them designed to do one thing and one thing well: they just
+receive `props` from their parent components and render! That's it. They are
+simple and beautiful and because they aren't doing much, because they are mostly
+stateless, they have a better chance of remaining blissfully bug free!
```
Great way to drive home how good presentational components are 🎉

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
-It's easy to go from here to building a `<Header />`, `<Footer />` etc.
+Now imagine that it's not just the `TextField` that our team has executed
+in this way, but say we've also defined a `<Header />` and a `<Footer />`, as
+well, of course, as other more unique and customized modules that are still
```
I feel like removing `, of course,` improves the readability a lot, the 'of course' breaks up the 'as well as' a little too much (might just be me)

ezmiller replied:
Yes definitely!

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+defines what we mean by a text input; it does so in a way that arguably covers
+the majority of use-cases we can imagine for a simple text input; and it provides
+this definition in one place that can be found and updated easily in the future.
+Win, win, win. Are we beginning to see the power off presentational components? 
```
'off' -> 'of'

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+    // Assume that we have the css classes defined elsewhere.
+    return <input
+             className="field field-light"
              onChange={this.props.onChange || noop}
```
Using `noop` makes sense here, but I think we should (very briefly) explain why we're doing this — some people might not know

ezmiller replied:
I think I'll just cut it. I think it distracts from what we are teaching at the moment.

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
   render() {
-    // assume that we have some of these classes defined
-    return <input className="field field-light"
+    // Assume that we have the css classes defined elsewhere.
```
I think this comment is a little redundant, we can assume that students expect this

ezmiller replied:
Okay.

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+no changeable state; and that only contains a render method. So, I think we can 
+safely say it's pretty well presentational.
+
+## Okay great, but when would we ever need such a simple component?
```
I feel like 'Okay' here causes repetitiveness (used in the last paragraph as well), can we rephrase this title a little bit and possibly make it a little shorter?

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+that does nothing but render a piece of our UI; that has no notion of how
+to fetch or reload the `message` data that it takes in as a `prop`; that has
+no changeable state; and that only contains a render method. So, I think we can 
+safely say it's pretty well presentational.
```
'pretty well presentational' -> should this be 'pretty much presentational' or just 'pretty presentational'? Obviously I'm not a native speaker so I'll defer to your judgment on this one 😄

ezmiller replied:
This is just a affectation. Was just playing with the language. But maybe it's too informal, and might be odd to non-native speakers. I'll fix.

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+Yep, this is true. Think about it. If a presentational component is simply
+a component that doesn't know anything about how to get the data it displays,
+and is mainly responsible for presentation, then you've been doing this from
+the beginning. A simple `HelloWorld` component, for example, is almost
```
Awesome job of easing them into this line of thinking, while simultaneously dropping the 💣

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+Okay so there's our pattern description. Now let's jump into some code and see
+how presentation components actually look in practice.
+
+## Surprise, You've already written presentational components!
```
I don't think we need to capitalize `You've` here

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+* they rarely have any internally changeable `state` properties;
+* and, they are best written as stateless functional components.
+
+Okay so there's our pattern description. Now let's jump into some code and see
```
Maybe add a comma after 'Okay' to improve the pacing a little bit?

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+* they probably only contain a render method;
+* they do not know how to load or alter the data that they render;
+* they rarely have any internally changeable `state` properties;
+* and, they are best written as stateless functional components.
```
Awesome description, really clear! 👍

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
+There is, however, a bit more thinking that we need to do here, just so we
+understand precisely what it means when we say that a component is of the type
+"presentational." Very often
+when we speak of classes or categories of things in world of programming, the types or
```
'in **the** world of programming', or just 'in programming'?

ezmiller replied:
I think it's better to keep this.

thomastuts commented on README.md:
```diff
@@ -1,55 +1,185 @@
-## Overview
+# What makes a component "presentational"?
+
+The answer to this question is, as you may have guessed, is that a presentational
```
I think we can leave out this first 'is'?

thomastuts commented:
@ezmiller This was a pleasure to read! This lesson explains the concept extremely well using great examples. Some nitpicks here and there, but the overall tone and content is absolutely great. 🔥

ezmiller committed:
Add md syntax identifier

ezmiller committed:
Fix indentation

ezmiller committed:
Improve phrasing

ezmiller committed:
Fix typo

ezmiller committed:
Remove duplice emdash

ezmiller committed:
Fix capitalization

ezmiller committed:
Improve phrasing

ezmiller committed:
Fix typo

ezmiller committed:
Remove noop

Needs explanation, but not worth getting into this since that
will distract from the point of the lesson.

ezmiller committed:
Remove unnecessary comment

ezmiller committed:
Make code styling consistent with other example

ezmiller committed:
Improve style

ezmiller committed:
Simplify conversational section header

ezmiller committed:
Formalize phrasing

ezmiller committed:
Add comma

ezmiller committed:
Fix typo

ezmiller committed:
Fix typo

ezmiller committed:
Add missing 'the'

ezmiller commented:
@thomastuts I've finished going through your valuable comments here. Let me know if you think it's ready to be merged.

thomastuts commented:
@ezmiller This looks great! :shipit:
